### PR TITLE
[SPARK-31438][CORE][SQL] Support JobCleaned Status in SparkListener

### PR DESCRIPTION
**Pull request by jackylee-ch**
### What changes were proposed in this pull request?
In Spark, we need do some hook after job cleaned, such as cleaning hive external temporary paths. This has already discussed in [GitHub Pull Request #28129](https://github.com/apache/spark/pull/28129).
The JobEnd Status is not suitable for this. As JobEnd is responsible for Job finished, once all result has generated, it should be finished. After finish, Scheduler will leave the still running tasks to be zombie tasks and delete abnormal tasks asynchronously.
Thus, we add JobCleaned Status to enable user to do some hook after all tasks cleaned in the job and add a JobCleanedListener to do some hooks.

### Does this PR introduce any user-facing change?
NO

### How was this patch tested?
run all tests

### CODE REVIEW

**Comment by stczwd** [comment]
cc @rdblue @cloud-fan

File: core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * JobCleanedHookListener is a basic job cleaned listener. It holds jobCleanedHooks for
+ * jobs and run cleaned hook after a job is cleaned.
+ */
+class JobCleanedHookListener extends SparkListener with Logging {
+  // use private val and synchronized to keep thread safe
+  private val jobCleanedHooks = new HashMap[Int, Int => Unit]()
+
+  def addCleanedHook(jobId: Int, fun: Int => Unit): Unit = synchronized{
```
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala:32]: `synchronized {`

File: core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    jobCleanedHooks.put(jobId, fun)
+  }
+
+  def deleteCleanedHook(jobId: Int): Unit = synchronized {
+    jobCleanedHooks.remove(jobId)
+  }
+
+  override def onJobCleaned(jobCleaned: SparkListenerJobCleaned): Unit = {
+    jobCleanedHooks.get(jobCleaned.jobId)
+      .foreach{function =>
```
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala:42]: `foreach { function`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 3]
```diff
@@ -99,6 +99,12 @@ case class InsertIntoHiveTable(
 
     try {
       processInsert(sparkSession, externalCatalog, hadoopConf, tableDesc, tmpLocation, child)
+    } catch {
+      case e: Throwable =>
+        // Make sure tmp path deleted while getting Exception before sc.runJob
+        deleteExternalTmpPath(hadoopConf)
+        throw new SparkException(
+          s"Failed inserting ubti table ${table.identifier.quotedString}", e)
```
Comment (xuanyuanking) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala:107]: ubti table?

File: core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import scala.collection.mutable.HashMap
+
+import org.apache.spark.internal.Logging
+
+/**
+ * JobCleanedHookListener is a basic job cleaned listener. It holds jobCleanedHooks for
+ * jobs and run cleaned hook after a job is cleaned.
+ */
+class JobCleanedHookListener extends SparkListener with Logging {
```
Comment (xuanyuanking) [inline core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala:28]: Maybe we need to add UT for the new class in SparkListenerSuite or somewhere else.

Comment (stczwd) [reply core/src/main/scala/org/apache/spark/scheduler/JobCleanedHookListener.scala:28]: Yes, I am doing it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala (hunk 3 above)
Comment (stczwd) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala:107]: sorry, wrong word

**Comment by stczwd** [comment]
cc @rdblue @cloud-fan @xuanyuanking

**Comment by stczwd** [comment]
kindly remind @rdblue @cloud-fan @xuanyuanking.

**Comment by cloud-fan** [comment]
> we need do some hook after job cleaned, such as cleaning hive external temporary paths.

Can you give more details about the use cases? There might be better ways to solve it.

**Comment by stczwd** [comment]
> Can you give more details about the use cases? There might be better ways to solve it.

@cloud-fan Thanks for your reply. Actually, this have been discussed in #28129. There are always some temporary directory left behind after Application Finished.

This is happened when we run `InsertIntoHiveTable` or `InsertIntoHiveDirCommand` with `spark.speculation=true`.
The execution environment has a slow response from the executor to driver, which causes some tasks to retry, and some tasks survive after the job ends. The surviving tasks continued writing result into temp dir, and this make temp dir uncleaned after application finished.

Here is the driver log for this.
```
2020-04-07 04:36:19 [dag-scheduler-event-loop]  INFO [DAGScheduler]: ResultStage 16 (sql at NativeMethodAccessorImpl.java:0) finished in 2.222 s
2020-04-07 04:36:19 [pool-3-thread-1]  INFO [DAGScheduler]: Job 2 finished: sql at NativeMethodAccessorImpl.java:0, took 23.883106 s
2020-04-07 04:36:19 [pool-3-thread-1]  INFO [FileFormatWriter]: Job null committed.
2020-04-07 04:36:19 [pool-3-thread-1]  INFO [FileFormatWriter]: Finished processing stats for job null.
2020-04-07 04:36:21 [task-result-getter-0]  WARN [TaskSetManager]: Lost task 752.0 in stage 16.0 (executor 42): TaskKilled (another attempt succeeded)
2020-04-07 04:36:21 [task-result-getter-0]  INFO [TaskSetManager]: Task 752.0 in stage 16.0 failed, but the task will not be re-executed (either because the task failed with a shuffle data fetch failure, so the previous stage needs to be re-run, or because a different copy of the task has already succeeded).
2020-04-07 04:36:21 [task-result-getter-3]  WARN [TaskSetManager]: Lost task 543.0 in stage 16.0 (executor 146): TaskKilled (another attempt succeeded)
2020-04-07 04:36:21 [task-result-getter-3]  INFO [TaskSetManager]: Task 543.0 in stage 16.0 failed, but the task will not be re-executed (either because the task failed with a shuffle data fetch failure, so the previous stage needs to be re-run, or because a different copy of the task has already succeeded).
2020-04-07 04:36:23 [task-result-getter-1]  WARN [TaskSetManager]: Lost task 143.0 in stage 16.0 (executor 146): TaskKilled (another attempt succeeded)
2020-04-07 04:36:23 [task-result-getter-1]  INFO [TaskSetManager]: Task 143.0 in stage 16.0 failed, but the task will not be re-executed (either because the task failed with a shuffle data fetch failure, so the previous stage needs to be re-run, or because a different copy of the task has already succeeded).
2020-04-07 04:36:28 [task-result-getter-1]  INFO [YarnClusterScheduler]: Removed TaskSet 16.0, whose tasks have all completed, from pool default
2020-04-07 04:36:28 [RPC-Handler-7]  INFO [SparkUI]: Stopped Spark web UI at http://192.168.1.125:48909
```

**Comment by cloud-fan** [comment]
is it possible that a task knows it fails to commit and clean up the files?

**Comment by stczwd** [comment]
> is it possible that a task knows it fails to commit and clean up the files?

Em, I have thought about this. 
This is a good way to remove extra output file and can also be used in `FileFormatWriter.write`.  However, it will still left temporary directory uncleaned, as each task only cleaned their output dir, none of them will clean the temporary dir.

**Comment by cloud-fan** [comment]
> as each task only cleaned their output dir, none of them will clean the temporary dir.

Can you give an example about the temp directory structure?

**Comment by stczwd** [comment]
> Can you give an example about the temp directory structure?

Here is one example for this.
```
.hive-staging_hive_2020-10-09_07-47-28_023_8877681090053776592-1/
└── -ext-10000
    └── _temporary
        └── 0
            ├── _temporary
            ├── task_20201009074730_0000_m_000000
            │   └── event_day=20180708
            │       └── part-00000-ffa4bddb-65a0-40a9-a94b-ba5eac475fb9.c000
            └── task_20201009074730_0000_m_000001
                └── event_day=20180708
                    └── part-00001-ffa4bddb-65a0-40a9-a94b-ba5eac475fb9.c000
```

**Comment by cloud-fan** [comment]
how about the task removing parent directory if it's empty?

**Comment by whcdjj** [comment]
Is there any progress on this issue? When I was insert into hive table with speculative enabled, I encountered the same problem.

**Pull request closed without merging**